handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -727,26 +665,63 @@ public final SSLEngineResult wrap(
                                 // So we attempt to drain the BIO buffer below, but if there is no data this condition
                                 // is undefined and we assume their is a fatal error with the openssl engine and close.
                                 // [1] https://www.openssl.org/docs/manmaster/ssl/SSL_write.html
-                                pendingNetResult = readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, status);
-                                return pendingNetResult != null ? pendingNetResult : NEED_WRAP_CLOSED;
+                                return newResult(NEED_WRAP, bytesConsumed, bytesProduced);
+
                             default:
                                 // Everything else is considered as error
                                 throw shutdownWithError("SSL_write");
                         }
                     }
                 }
-            }
-            // We need to check if pendingWrittenBytesInBIO was checked yet, as we may not checked if the srcs was
-            // empty, or only contained empty buffers.
-            if (bytesConsumed == 0) {
-                SSLEngineResult pendingNetResult = readPendingBytesFromBIO(dst, 0, bytesProduced, status);
-                if (pendingNetResult != null) {
-                    return pendingNetResult;
+                return newResultMayFinishHandshake(status, bytesConsumed, bytesProduced);
+            } finally {
+                SSL.bioClearByteBuffer(networkBIO);
+                if (bioReadCopyBuf == null) {
+                    dst.position(dst.position() + bytesProduced);
+                } else {
+                    assert bioReadCopyBuf.readableBytes() <= dst.remaining() : "The destination buffer " + dst +
+                            " didn't have enough remaining space to hold the encrypted content in " + bioReadCopyBuf;
+                    dst.put(bioReadCopyBuf.internalNioBuffer(bioReadCopyBuf.readerIndex(),
+                                                      bioReadCopyBuf.readerIndex() + bytesProduced));
+                    bioReadCopyBuf.release();
```

Comment:
I don't think this is necessary. The assert signifies a programatic error and was added to aid development and unit tests. Do you have a suggestion which doesn't introduce runtime checks or additional try/catch?